mea/src/broadcast/tests.rs

```diff
@@ -0,0 +1,214 @@
+        _ => panic!("Expected Lagged error"),
+    }
+
+    assert_eq!(rx.recv().await, Ok(3));
+    assert_eq!(rx.recv().await, Ok(4));
+}
+
+#[tokio::test]
+async fn test_broadcast_closed() {
+    let (tx, mut rx) = channel::<i32>(10);
+    drop(tx);
+    assert_eq!(rx.recv().await, Err(RecvError::Disconnected));
+}
+
+#[tokio::test]
+async fn test_wrap_around() {
+    let (tx, mut rx) = channel(2);
+    for i in 0..10 {
+        tx.send(i);
+        if i >= 2 {
+            if let Ok(v) = rx.recv().await {
+                assert_eq!(v, i);
+            }
+        }
+    }
+}
+
+#[tokio::test]
+async fn test_wait_mechanism() {
+    let (tx, mut rx) = channel(10);
+
+    let handle = tokio::spawn(async move { rx.recv().await });
+
+    tokio::time::sleep(std::time::Duration::from_millis(10)).await;
+    tx.send(42);
+
+    assert_eq!(handle.await.unwrap(), Ok(42));
+}
+
+#[tokio::test]
+async fn test_subscribe() {
+    let (tx, _rx) = channel::<i32>(10);
+    let mut rx = tx.subscribe();
+
+    tx.send(100);
+    assert_eq!(rx.recv().await, Ok(100));
+}
+
+#[tokio::test]
+async fn test_resubscribe() {
+    let (tx, mut rx) = channel(2);
+
+    tx.send(1);
+    tx.send(2);
+
+    let mut rx2 = rx.resubscribe();
+
+    // rx sees 1, 2
+    // rx2 sees nothing yet (starts at tail=2)
+
+    tx.send(3);
+
+    match rx.recv().await {
+        Err(RecvError::Lagged(n)) => assert_eq!(n, 1),
+        _ => panic!("Expected Lagged error"),
+    }
+    assert_eq!(rx.recv().await, Ok(2));
+    assert_eq!(rx2.recv().await, Ok(3));
+}
+
+#[tokio::test]
+async fn test_overflow() {
+    let (tx, mut rx) = channel(4);
+    let mut rx2 = rx.clone();
+
+    let boundary = usize::MAX - 2;
+    tx.shared.tail_cnt.store(boundary, Ordering::SeqCst);
+    rx.head = boundary;
+
+    tx.send(1);
+    assert_eq!(rx.recv().await, Ok(1));
+
+    tx.send(2);
+    tx.send(3);
+    tx.send(4);
+    tx.send(5);
+    tx.send(6);
+    tx.send(7);
+    tx.send(8);
+
+    assert_eq!(rx.recv().await, Err(RecvError::Lagged(3)));
+    assert_eq!(rx.recv().await, Ok(5));
+    assert_eq!(rx.recv().await, Ok(6));
+    assert_eq!(rx.recv().await, Ok(7));
+    assert_eq!(rx.recv().await, Ok(8));
+
+    assert_eq!(rx2.recv().await, Err(RecvError::Lagged(1)));
+    assert_eq!(rx2.recv().await, Ok(5));
+    assert_eq!(rx2.recv().await, Ok(6));
+    assert_eq!(rx2.recv().await, Ok(7));
+    assert_eq!(rx2.recv().await, Ok(8));
+}
+
+#[tokio::test]
+async fn test_overflow_wrapping() {
+    let (tx, mut rx) = channel(4);
+    let mut rx2 = rx.clone();
+
+    let boundary = usize::MAX - 2;
+    tx.shared.tail_cnt.store(boundary, Ordering::SeqCst);
+    rx.head = boundary;
+
+    tx.send(1);
+    assert_eq!(rx.recv().await, Ok(1));
+
+    tx.send(2);
+    tx.send(3);
+    tx.send(4);
+    tx.send(5);
+
+    assert_eq!(rx.recv().await, Ok(2));
+    // FIXME: wrapping just hit the head, but this should be lagged
+    assert_eq!(rx2.recv().await, Ok(4));
```

Comment:
We need to figure out what to do with this case. May or may not be able to work it out.

Perhaps a better "version" strategy and a proper forward solution.